shared/actions/search.js

```diff
@@ -1,102 +1,151 @@
+  const results: Array<SearchResult> = filterNull(rresults.map(rr => {
+    const extraInfo = parseExtraInfo(platform, rr)
+    const serviceName = rr.service && rr.service.service_name && capitalize(rr.service.service_name)
 
-  const doRPC = (...args) => new Promise((resolve, reject) => {
-    // TODO think about using rpc
-    engine.rpcUnchecked(...args, (error, results) => {
-      if (badNonce()) { return }
-      if (error) { throw new Error(error) }
-      if (results) {
-        dispatch(loadSummaries(results.map(r => r.uid)))
+    if (platform === 'Keybase') {
+      return {
+        service: 'keybase',
+        username: rr.keybase.username,
+        isFollowing: rr.keybase.is_followee,
+        extraInfo,
       }
-      resolve(results || [])
-    })
-  })
+    } else if (serviceName) {
+      return {
+        service: 'external',
+        icon: platformToLogo32(serviceName),
+        username: rr.service.username,
+        extraInfo,
+      }
+    } else {
+      return null
+    }
+  }))
 
-  Promise.all([
-    doRPC('user.listTracking', {filter: term}, {}).then(results => {
-      return results.map(r => ({uid: r.uid, username: r.username, tracking: true}))
-    }),
-    doRPC('user.search', {query: term}, {}).then(results => {
-      return results.map(r => ({uid: r.uid, username: r.username}))
-    }),
-  ])
-    .then(results => {
-      const trackingUsernames = new Set(results[0].map(u => u.uid))
-      dispatch({
-        type: Constants.searchResults,
-        payload: {
-          base,
-          results: results[0].concat(results[1].filter(r => !trackingUsernames.has(r.uid))),
-        },
-      })
-    })
-    .catch(err => dispatch({
-      type: Constants.searchResults,
-      payload: {
-        base,
-        error: err,
-      },
-    }))
-}, 150)
+  return {
+    type: Constants.results,
+    payload: {term, results},
+  }
+}
 
-export function submitSearch (base, term) {
-  return (dispatch, getState) => {
-    if (term === '') {
-      // Clears any existing search results
-      return dispatch(initSearch(base))
-    }
+export function search (term: string, platform: SearchPlatforms = 'Keybase') : TypedAsyncAction<Search | Results> {
+  return dispatch => {
     dispatch({
-      type: Constants.searchTerm,
+      type: Constants.search,
       payload: {
-        base,
         term,
+        error: false,
       },
     })
-    submitSearchDebounced(base, term, dispatch, getState)
+
+    // TODO daemon rpc, for now api hit
+    // const params: UserSearchRpc = {
+      // method: 'user.search',
+      // param: {
+        // query: term
+      // },
+      // incomingCallMap: {},
+      // callback: (error: ?any, uresults: UserSearchResult) => {
+        // if (error) {
+          // console.log('Error searching. Not handling this error')
+        // } else {
+          // dispatch(results(term, uresults))
+        // }
+      // }
+    // }
+
+    // engine.rpc(params)
+    const service = {
```

Comment:
It my be worthwhile to type the keys in this to ensure that they are ServicePlatforms. Is there someway for flow to ensure that all cases are covered (so that we can't leave out a platform)?